### PR TITLE
Support `has` in rename refactoring

### PR DESCRIPTION
**Pull request by toinehartman**
This PR implements support for renaming field names used in `<Expression> has <Name>`. Since this uses a dynamic check and the name is not statically resolved to any definition, this requires user interaction to preserve semantics.
This PR also adds framework support user-confirmable rename suggestions.

### CODE REVIEW

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java [hunk 1]
```diff
@@ -287,8 +289,12 @@ public CompletableFuture<WorkspaceEdit> rename(RenameParams params) {
             .thenApply(Versioned::get)
             .handle((t, r) -> (t == null ? (file.getMostRecentTree().get()) : t))
             .thenCompose(tr -> rascalServices.getRename(tr, params.getPosition(), workspaceFolders, facts::getPathConfig, params.getNewName(), columns).get())
-            .thenApply(c -> new WorkspaceEdit(DocumentChanges.translateDocumentChanges(this, c)))
-            ;
+            .thenApply(t -> {
+                WorkspaceEdit wsEdit = new WorkspaceEdit();
+                wsEdit.setDocumentChanges(DocumentChanges.translateDocumentChanges(this, (IList) t.get(0)));
+                wsEdit.setChangeAnnotations(DocumentChanges.translateChangeAnnotations((IMap) t.get(1)));
```
[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:295]: 1. Mark changes as possibly requiring user interaction before being applied.
2. Provide a description.

In the UI, that might look like this:
afbeelding

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/Rename.rsc [hunk 2]
```diff
@@ -227,17 +229,26 @@ Maybe[loc] rascalLocationOfName(Nonterminal nt) = just(nt.src);
 Maybe[loc] rascalLocationOfName(NonterminalLabel l) = just(l.src);
 default Maybe[loc] rascalLocationOfName(Tree t) = nothing();
 
-private tuple[set[IllegalRenameReason] reasons, list[TextEdit] edits] computeTextEdits(WorkspaceInfo ws, start[Module] m, set[loc] defs, set[loc] uses, str name) {
-    if (reasons := rascalCollectIllegalRenames(ws, m, defs, uses, name), reasons != {}) {
+private tuple[set[IllegalRenameReason] reasons, list[TextEdit] edits] computeTextEdits(WorkspaceInfo ws, start[Module] m, set[RenameLocation] defs, set[RenameLocation] uses, str name, ChangeAnnotationRegister registerChangeAnnotation) {
```
[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/Rename.rsc]: Correct! Missed the typechecker warning about that. Fixed now.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:305]: I was a bit surprised that the edits that we produced before - that were applied without user interaction - are now suddenly also shown in this list, and can be unconfirmed by the user. I'll experiment with this and see if we can prevent that. Otherwise, we can attach a description to them as well, to persuade the user not to touch them.

[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:305]: If we send any edits with `needsConfirmation = true` (like right-hand sides of `has`), all edits will [always appear](https://github.com/microsoft/vscode/blob/6e3cbfd138073c018087744c5597209f61f0ad18/src/vs/workbench/contrib/bulkEdit/browser/bulkEditService.ts#L190) in this 'refactor preview'. I do not see any way to lock the non-confirmable edits to prevent them being unchecked by the user.

Therefore, I followed your suggestion and added some descriptions to the headers. What do you think of the following @DavyLandman ?

afbeelding

[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:305]: If there are no to-be-confirmed instances (currently: no uses of `has <name-under-cursor>`), the preview will still be skipped.

[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:305]: Yes, much better, how about rewording it:

> Due to the dynamic behavior of these uses, we are not sure if they should also be renamed, please review these changes.

and

> We're positive these edits are necessary, they are only shown here to get a full overview, we advice not to disable them.

[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:305]: good enough :+1:

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java [hunk 3]
```diff
@@ -231,7 +232,7 @@ public InterruptibleFuture<IList> getRename(ITree module, Position cursor, Set<I
                 }
                 throw e;
             }
-        }, VF.list(), exec, false, client);
+        }, VF.tuple(), exec, false, client);
```
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:235]: this empty tupple will fail if we do `.get(0)` on it.

Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:235]: perhaps we should generate a tuple with 2 empty lists?

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+POSSIBILITY OF SUCH DAMAGE.
+}
+module lang::rascal::lsp::refactor::TextEdits
+
+extend analysis::diff::edits::TextEdits;
+
+alias ChangeAnnotationId = str;
+
+data ChangeAnnotation
+    = changeAnnotation(str label, str description, bool needsConfirmation)
```
Comment (DavyLandman) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc:34]: how come we don't add `bool needsConfirmation` as an ADT to the `TextEdit`? which would be a feature we could then also consider porting over to rascal?

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc [hunk 5]
```diff
@@ -66,7 +69,8 @@ data Cursor
 
 alias MayOverloadFun = bool(set[loc] defs, map[loc, Define] defines);
 alias FileRenamesF = rel[loc old, loc new](str newName);
-alias DefsUsesRenames = tuple[set[loc] defs, set[loc] uses, FileRenamesF renames];
+data RenameLocation = rl(loc l, Maybe[ChangeAnnotationId] annotation = nothing());
```
Comment (DavyLandman) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:72]: smells like a tuple. is there a reason why it's an adt? 

```
tuple[rel[loc, Maybe[ChangeAnnotationId]] defs...
```

would have been possible?

it reads to me as a bit messy to annotate the locations like this.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc [hunk 6]
```diff
@@ -531,10 +539,23 @@ DefsUsesRenames rascalGetDefsUses(WorkspaceInfo ws, cursor(typeParam(), cursorLo
...
+    return <fromLocs(defs), fromLocs(useDefs - defs), NO_RENAMES>;
 }
 
-DefsUsesRenames rascalGetDefsUses(WorkspaceInfo ws, cursor(dataField(loc adtLoc, AType fieldType), cursorLoc, cursorName), MayOverloadFun mayOverloadF) {
+private str describeFact(just(AType tp)) = "type \'<prettyAType(tp)>\'";
+private str describeFact(nothing()) = "unknown type";
+
+set[RenameLocation] rascalGetHasUses(WorkspaceInfo ws, set[loc] defs, str cursorName, ChangeAnnotationRegister registerChangeAnnotation) {
+    return {
+        rl(name.src, annotation = just(registerChangeAnnotation("Use of `has <cursorName>` on value of <describeFact(getFact(ws, e.src))>", "Due to the dynamic nature of these names, it cannot be ensured that they need to be renamed. Please review these suggested changes.", true)))
```
Comment (DavyLandman) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:550]: > it cannot be ensured that they need to be renamed

How about:

> Due to the dynamic natures of these names, please review these suggested changes.

**Review by DavyLandman — APPROVED** [review]
Some small comments and questions.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java (hunk 3 above)
Comment (DavyLandman) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:235]: yes, I would expect so.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc (hunk 5 above)
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc]: You're right. It's just a tuple. Refactored!

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc (hunk 6 above)
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc]: Rewritten

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc (hunk 4 above)
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc:34]: Done.

**Review by DavyLandman — APPROVED** [review]
LGTM, 2 small remarks. Merge when you feel you've dealt with them.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc [hunk 7]
```diff
@@ -310,6 +315,11 @@ bool rascalIsCollectionType(AType at) = at is arel || at is alrel || at is atupl
 bool rascalIsConstructorType(AType at) = at is acons;
 bool rascalIsDataType(AType at) = at is aadt;
 
+bool rascalIsDataTypeLike(dataId()) = true;
```
[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:318]: Oh this is interesting. There are no tests for renaming these 'nodes' at all yet (I was not even aware of their existence). Let me look into this / add some tests.

[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:318]: I'm not sure how much the typechecker even support field deref on raw node values.

[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:318]: Looked into this for a bit. Currently, there is not support for nodes at all (since we assume that field LHSs refer to ADT or constructor types). We can support this, but I propose to do this in a separate PR, since I foresee quite some changes.
https://github.com/usethesource/rascal-language-servers/issues/510

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+module lang::rascal::lsp::refactor::TextEdits
+
+extend analysis::diff::edits::TextEdits;
+
+alias ChangeAnnotationId = str;
```
[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc:31]: Not really needed, but I think this is a way to communicate the intent of the string, instead of returning a bunch of strings where one is the id, another is the label and a third is the description.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc [hunk 9]
```diff
@@ -423,70 +429,72 @@ private set[loc] rascalGetExceptUses(WorkspaceInfo ws, set[loc] defs) {
...
 
-    set[loc] uses = {};
+    set[RenameLocation] uses = {};
     for (Define d: <_, consName, _, constructorId(), _, defType(acons(aadt(_, _, _), _, _))> <- constructorDefs) {
         // Find all neighbouring pairs of facts where an except for `cursorName` exists only in the latter
         for (
-            [ _*
-            , <l1, at1: !/\a-except(consName)>
+            [ *_
+            , <_, _: !/\a-except(consName)>
```
[the remark this answers is not in the document]
Comment (toinehartman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:437]: Right, missed that when dropping the names. Fixed.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc (hunk 8 above)
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/TextEdits.rsc:31]: It's not a thing we normally do in the api, but let's tackle that when we get to make this part of rascal-lsp.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc (hunk 7 above)
[the remark this answers is not in the document]
Comment (DavyLandman) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc:318]: Agreed, better to pick up later.
